File: packages/gavio-js/tests/unit/stream-buffer.test.ts
import { describe, it, expect } from 'vitest'
import { createHash } from 'node:crypto'
import { Gateway } from '../../src/gateway.js'
import { StreamBuffer } from '../../src/interceptors/reliability/index.js'
import {
  auditInterceptor,
  type AuditSink,
  type AuditRecord,
} from '../../src/interceptors/audit/index.js'
import type { Interceptor } from '../../src/interceptors/base.js'
import type { InterceptorContext } from '../../src/context.js'
import type { GavioResponse } from '../../src/response.js'

class Collector implements AuditSink {
  records: AuditRecord[] = []
  async write(record: AuditRecord): Promise<void> {
    this.records.push(record)
  }
}

/** A post-interceptor that rewrites the response content. */
const shout: Interceptor = {
  name: 'shout',
  async after(response: GavioResponse, _ctx: InterceptorContext): Promise<GavioResponse> {
    response.content = response.content.toUpperCase()
    return response
  },
}

const collect = async (gen: AsyncGenerator<string>): Promise<string> => {
  let out = ''
  for await (const chunk of gen) out += chunk
  return out
}

describe('streaming reliability / StreamBuffer (F-REL-06)', () => {
  it('StreamBuffer accumulates chunks', () => {
    const buf = new StreamBuffer()
    expect(buf.text()).toBe('')
    expect(buf.length).toBe(0)
    buf.append('ab')
    buf.append('cd')
    expect(buf.text()).toBe('abcd')
    expect(buf.length).toBe(4)
  })

  it('emits the buffered content', async () => {
    const gw = new Gateway({ devMode: true })
    const full = await collect(gw.stream({ messages: [{ role: 'user', content: 'hi there' }] }))
    expect(full.trim()).toBe('[mock reply] hi there')
  })

  it('runs post-interceptors on the full buffered response', async () => {
    const sink = new Collector()
    const gw = new Gateway({ devMode: true }).use(auditInterceptor({ sink }))
    const full = await collect(gw.stream({ messages: [{ role: 'user', content: 'hello world' }] }))

    expect(sink.records).toHaveLength(1)
    const expected = createHash('sha256').update(full, 'utf-8').digest('hex')
    expect(sink.records[0]!.responseHash).toBe(expected)
    expect(sink.records[0]!.interceptorsFired).toContain('audit')
  })

  it('a post-interceptor rewrite is visible to the caller', async () => {
    const gw = new Gateway({ devMode: true }).use(shout)
    const full = await collect(gw.stream({ messages: [{ role: 'user', content: 'quiet' }] }))
    expect(full).toBe(full.toUpperCase())
    expect(full).toContain('[MOCK REPLY] QUIET')
  })
})

File: packages/gavio-java/gavio-core/src/test/java/io/gavio/StreamBufferTest.java
package io.gavio;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import io.gavio.interceptors.Interceptor;
import io.gavio.interceptors.InterceptorContext;
import io.gavio.providers.MockProvider;
import io.gavio.providers.StreamBuffer;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicReference;
import org.junit.jupiter.api.Test;

/** Streaming reliability / StreamBuffer (F-REL-06). */
class StreamBufferTest {

    /** A post-interceptor that rewrites the response content. */
    private static final class Shout implements Interceptor {
        @Override
        public String name() {
            return "shout";
        }

        @Override
        public CompletableFuture<GavioResponse> after(GavioResponse response, InterceptorContext ctx) {
            return CompletableFuture.completedFuture(response.withContent(response.content().toUpperCase()));
        }
    }

    /** A post-interceptor that records the response content it observed. */
    private static final class Capture implements Interceptor {
        final AtomicReference<String> seen = new AtomicReference<>();

        @Override
        public String name() {
            return "capture";
        }

        @Override
        public CompletableFuture<GavioResponse> after(GavioResponse response, InterceptorContext ctx) {
            seen.set(response.content());
            return CompletableFuture.completedFuture(response);
        }
    }

    private static String drain(java.util.concurrent.Flow.Publisher<String> publisher) {
        return StreamBuffer.collect(publisher).join().text();
    }

    private static GavioRequest req(String content) {
        return GavioRequest.builder().message("user", content).model("mock").build();
    }

    @Test
    void streamBufferAccumulates() {
        StreamBuffer buf = new StreamBuffer();
        assertEquals("", buf.text());
        assertEquals(0, buf.length());
        buf.append("ab");
        buf.append("cd");
        assertEquals("abcd", buf.text());
        assertEquals(4, buf.length());
    }

    @Test
    void collectAssemblesProviderStream() {
        String text = StreamBuffer.collect(new MockProvider().stream(req("hi there"))).join().text();
        assertEquals("[mock reply] hi there", text.trim());
    }

    @Test
    void gatewayStreamEmitsBufferedContent() {
        Gateway gw = Gateway.builder().devMode(true).build();
        String full = drain(gw.stream(req("hi there")));
        assertEquals("[mock reply] hi there", full.trim());
    }

    @Test
    void postInterceptorSeesFullBufferedResponse() {
        Capture capture = new Capture();
        Gateway gw = Gateway.builder().devMode(true).use(capture).build();
        String full = drain(gw.stream(req("hello world")));
        assertEquals(full, capture.seen.get());
        assertTrue(full.contains("hello world"));
    }

    @Test
    void postInterceptorRewriteVisibleToCaller() {
        Gateway gw = Gateway.builder().devMode(true).use(new Shout()).build();
        String full = drain(gw.stream(req("quiet")));
        assertEquals(full.toUpperCase(), full);
        assertTrue(full.contains("[MOCK REPLY] QUIET"));
    }
}
